scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/license/LicenseSynchronization.java

```diff
@@ -0,0 +1,218 @@
+
+            recordManager.commit();
+            if (!firstAttempt) {
+                logger.info("Loading of job database successful after clean.");
+            }
+
+        } catch (Throwable e) {
+            if (firstAttempt) {
+                logger.error("Error occurred when loading the database " + storeFile.getAbsolutePath() +
+                             ", now cleaning it and retrying.", e);
+                cleanDataBase();
+                initDBMaps(false);
+            } else {
+                close();
+                throw new IllegalStateException("Error when loading database (even after cleaning it): " +
+                                                storeFile.getAbsolutePath(), e);
+            }
+        }
+    }
+
+    void addSoftware(String software, int nbTokens) {
+        persistedJobsLicenses.put(software, new LinkedBlockingQueue<String>(nbTokens));
+        persistedTasksLicenses.put(software, new LinkedBlockingQueue<String>(nbTokens));
+        remainingTokens.put(software, nbTokens);
+    }
+
+    void persist() {
+        try {
+            recordManager.commit();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    LinkedBlockingQueue<String> getPersistedJobsLicense(String license) {
+        return persistedJobsLicenses.get(license);
+    }
+
+    boolean containsJobId(String license, String jobId) {
+        return persistedJobsLicenses.get(license).stream().anyMatch(jid -> jid.equals(jobId));
+    }
+
+    void addJobToLicense(String license, String jobId) {
+        persistedJobsLicenses.get(license).add(jobId);
+        remainingTokens.put(license, remainingTokens.get(license) - 1);
+    }
+
+    void markJobLicenseChange(String license) {
+        // Record Manager mark as dirty (uncommited) entries which have be modified via a put call
+        // Thus, such operation as persistedJobsLicenses.get(license).dosomething() will not be committed
+        // by the following trick, we mark the entry as dirty and commit
+        persistedJobsLicenses.put(license, persistedJobsLicenses.get(license));
+    }
+
+    LinkedBlockingQueue<String> getPersistedTasksLicense(String license) {
+        return persistedTasksLicenses.get(license);
+    }
+
+    void addTaskToLicense(String license, String taskId) {
+        persistedTasksLicenses.get(license).add(taskId);
+        remainingTokens.put(license, remainingTokens.get(license) - 1);
+    }
+
+    void markTaskLicenseChange(String license) {
+        // Record Manager mark as dirty (uncommited) entries which have be modified via a put call
+        // Thus, such operation as persistedTasksLicenses.get(license).dosomething() will not be committed
+        // by the following trick, we mark the entry as dirty and commit
+        persistedTasksLicenses.put(license, persistedTasksLicenses.get(license));
+    }
+
+    int getRemainingTokens(String license) {
```

Comment:
```suggestion
    Integer getRemainingTokens(String license) {
```

Reply:
Because if licence is not there, you will get NPE.

Reply:
I will keep int but: will add a log + re throws the exception.